hugr/src/algorithm/merge_bbs.rs

```diff
@@ -0,0 +1,399 @@
+                .unwrap();
+            let entry_nop = nops[entry_nop_idx];
+            nops.remove(entry_nop_idx);
+            let [n_op2] = nops.try_into().unwrap();
+            assert_eq!(h.get_parent(n_op2), Some(no_b2));
+            (entry_nop, no_b2)
+        };
+        assert_eq!(h.get_parent(entry_nop), Some(entry));
+        assert_eq!(
+            h.output_neighbours(entry).collect::<HashSet<_>>(),
+            HashSet::from([expected_backedge_target, exit])
+        );
+        // And the Noop in the entry block is consumed by the custom Test op
+        let tst = find_unique(h.nodes(), |n| {
+            matches!(h.get_optype(*n), OpType::CustomOp(_))
+        });
+        assert_eq!(h.get_parent(tst), Some(entry));
+        assert_eq!(
+            h.output_neighbours(entry_nop).collect::<Vec<_>>(),
+            vec![tst]
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn triple_with_permute() -> Result<(), Box<dyn std::error::Error>> {
+        // Blocks are just BB1 -> BB2 -> BB3 --> Exit.
+        // CFG Normalization would move everything outside the CFG and elide the CFG altogether,
+        // but this is an easy-to-construct test of merge-basic-blocks only (no CFG normalization).
+        let e = extension();
+        let tst_op: OpType = e
+            .instantiate_extension_op("Test", &[], &PRELUDE_REGISTRY)?
+            .into();
+        let [res_t] = tst_op
+            .dataflow_signature()
+            .unwrap()
+            .output
+            .into_owned()
+            .try_into()
+            .unwrap();
+        let mut h = CFGBuilder::new(
+            FunctionType::new(QB_T, res_t.clone())
+                .with_extension_delta(ExtensionSet::singleton(&PRELUDE_ID)),
+        )?;
+        let mut bb1 = h.entry_builder(
+            vec![type_row![]],
+            type_row![USIZE_T, QB_T],
+            ExtensionSet::singleton(&PRELUDE_ID),
+        )?;
+        let [inw] = bb1.input_wires_arr();
+        let load_cst = bb1.add_load_value(ConstUsize::new(1));
+        let pred = lifted_unary_unit_sum(&mut bb1);
+        let bb1 = bb1.finish_with_outputs(pred, [load_cst, inw])?;
+
+        let mut bb2 = h.block_builder(
+            type_row![USIZE_T, QB_T],
+            vec![type_row![]],
+            ExtensionSet::new(),
+            type_row![QB_T, USIZE_T],
+        )?;
+        let [u, q] = bb2.input_wires_arr();
+        let pred = lifted_unary_unit_sum(&mut bb2);
+        let bb2 = bb2.finish_with_outputs(pred, [q, u])?;
+
+        let mut bb3 = h.block_builder(
+            type_row![QB_T, USIZE_T],
+            vec![type_row![]],
+            ExtensionSet::new(),
+            res_t.clone().into(),
+        )?;
+        let [q, u] = bb3.input_wires_arr();
+        let tst = bb3.add_dataflow_op(tst_op, [q, u])?;
+        let pred = lifted_unary_unit_sum(&mut bb3);
+        let bb3 = bb3.finish_with_outputs(pred, tst.outputs())?;
+        println!("ALAN bb1 {bb1:?} bb2 {bb2:?} bb3 {bb3:?}");
```

Comment:
Looks like a debug line left in by mistake?

Reply:
Oops, thanks!